### PR TITLE
Habit event testing, feedback, bug fixing and UI updates + image upload/display

### LINKED ISSUE #158: Convert addEventFragment to an Activity
This issue is related to user story #31 
Use a new activity for adding habit event instead of using a fragment.

### LINKED ISSUE #159: Launch habit event list activity from a habit
This issue is related to #31 

According to what the TA said, habit event list should not be on its own. The user should be able to open a detail view of a habit then click a button to display all habit events related the selected habit.

It involves the following changes: 
* add a 'see events' button in habit detail view
* pass a habit object from habit detail view to event list and display all events related to the selected habit.

### LINKED ISSUE #164: Add/Edit/Delete Habit Event Feedback
This issue is related to user story #45 #42 # 39 and #31 

User should get feedbacks after they add/edit/delete a habit event

Task list:
- [x] Use toast to display feedback messages.

### LINKED ISSUE #165: Use custom intent testing rule to prevent from testing failure
This issue is related to #45 #42 #39 and #31

There are several testing failure after a small modification on display user name. This failure is caused by activity is started before mock user is added.
A custom intent testing rule is used to add mock user before start an activity.

Task list:
- [x] Re-write test cases by using custom intent testing rule.
- [x] Add timeouts in intent testing to prevent test failure caused by slow computer.

### LINKED ISSUE #166: Combine image upload/display section (created by Jie) with newest habit event UI
This issue is related to user story #80 

Combine Jie's work with newest habit event UI.

Task list:

- [x] merge image upload/display codes with UI codes.

### LINKED ISSUE #167: Replace getAllHabitEvent by a snapshot listener
This issue is related to user story #39 

The upload/display image function created by Jie is asynchronous and slow. There is a chance that getAllHabitEvent function failed to fetch all habit events from firestore database because the image uploading is not finished yet.
Using snapshot listener could solve this problem easily.

Task list:
- [x] replace getAllHabitEvents function by a custom snapshot listener.

### PR DESCRIPTION
**Pull request by johnzhang27**
This PR is related the following issues:
- #159 Habit event list now is launched from a given habit detail view and it only display the habit event associated with the selected habit.
- #158 Since we need more space to display details of a habit event (image, comment), the fragment has been converted to an activity.
- #164 Use toast to display feedback messages after user add/edit/delete a habit event.
- #165 Use custom intent testing rule created by @zarifmahfuz to re-write intent tests related to habit event.
- #166 Merged image upload/display function created by @Jie-F to the newest habit event UI. @Jie-F needs to add some javadoc for image upload/display functions.
- #167 Use a snapshot listener in HabitEventList Activity instead of getAllHabitEvents function. Have to do this because of an asynchronous error.

close #159 
close #158 
close #164 
close #165 
close #166 
close #167

### CODE REVIEW

**Review by Jie-F — APPROVED** [review]
Lots of good stuff, nice work!

File: code/HabitTracker/app/src/main/java/com/example/habittracker/activities/eventlist/CustomList.java [hunk 1]
```diff
@@ -62,6 +69,8 @@ public CustomList(Context context, ArrayList<HabitEvent> events){
         super(context,0, events);
         this.events = events;
         this.context = context;
+        mStorage = FirebaseStorage.getInstance();
+        mDatabaseRef = FirebaseDatabase.getInstance().getReference("uploads");
```
Comment (PaoVal) [inline code/HabitTracker/app/src/main/java/com/example/habittracker/activities/eventlist/CustomList.java:73]: hey so. I noticed this "uploads" reference is not documented anywhere.
At least it's not in the schema. Maybe this is something that should be added but idk if this is more of a @Jie-F thing

File: code/HabitTracker/app/src/main/java/com/example/habittracker/activities/eventlist/EventDetailActivity.java [hunk 2]
```diff
@@ -5,18 +5,26 @@
 
 import static java.security.AccessController.getContext;
 
+import android.content.Context;
```
Comment (PaoVal) [inline code/HabitTracker/app/src/main/java/com/example/habittracker/activities/eventlist/EventDetailActivity.java:8]: There's missing javadoc for this entire file.

File: code/HabitTracker/app/src/test/java/com/example/habittracker/DatabaseManagerTest.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
+//package com.example.habittracker;
+//
```
Comment (PaoVal) [inline code/HabitTracker/app/src/test/java/com/example/habittracker/DatabaseManagerTest.java:2]: Mmm ok so I see a lot of commented out test code here. What happened? Is it failing? Do we need to make an issue for someone to fix this?

File: code/HabitTracker/app/src/test/java/com/example/habittracker/ProgressUtilTest.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+import org.junit.jupiter.api.Test;
+
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.GregorianCalendar;
+
+public class ProgressUtilTest {
+
+    /**I'll leave test case for getOverallProgress function to Nick*/
```
Comment (PaoVal) [inline code/HabitTracker/app/src/test/java/com/example/habittracker/ProgressUtilTest.java:15]: @NickSNFK - thought I'd bring this to your attention.

**Comment by PaoVal** [comment]
Looks generally good. I tested it and ran into no problems. So I'd say to go ahead. Please just push the UI changes you shared with the team before. And then I'd say good to go! (Only actual thing is to please comment the event detail activity. I don't see any javadoc there, or it might also be GitHub bugging out again)

File: code/HabitTracker/app/src/main/java/com/example/habittracker/activities/eventlist/EventDetailActivity.java (hunk 2 above)
Comment (johnzhang27) [reply code/HabitTracker/app/src/main/java/com/example/habittracker/activities/eventlist/EventDetailActivity.java:8]: @Jie-F Could you add javadoc later?

File: code/HabitTracker/app/src/test/java/com/example/habittracker/DatabaseManagerTest.java (hunk 3 above)
Comment (johnzhang27) [reply code/HabitTracker/app/src/test/java/com/example/habittracker/DatabaseManagerTest.java:2]: We are having a connection error to FirestoreApp and that's why the entire script is commented out. We might want to delete this script if the error can't be resolved in time.

**Review by PaoVal — APPROVED** [review]

**Comment by johnzhang27** [comment]
> Looks generally good. I tested it and ran into no problems. So I'd say to go ahead. Please just push the UI changes you shared with the team before. And then I'd say good to go! (Only actual thing is to please comment the event detail activity. I don't see any javadoc there, or it might also be GitHub bugging out again)

New changes has been pushed, check commit #4d6dc5e

**Pull request merged**